File: fe/types/quiz.ts
export interface Quiz {
    id: number;
    module_id: number;
    title: string;
    description: string;
    passing_score: number;
    total_questions: number;
    is_active: boolean;
    created_at: string;
    questions: QuizQuestion[];
}

export interface QuizQuestion {
    id: number;
    quiz_id: number;
    question_text: string;
    question_type: "multiple_choice" | "true_false" | "short_answer";
    options?: string[];
    correct_answer?: string;
    explanation?: string;
    points: number;
    order_index: number;
}

export interface QuizForTaking {
    id: number;
    module_id: number;
    title: string;
    description: string;
    passing_score: number;
    total_questions: number;
    estimated_completion_time: number;
    questions: QuizQuestionForTaking[];
}

export interface QuizQuestionForTaking {
    id: number;
    question_text: string;
    question_type: "multiple_choice" | "true_false" | "short_answer";
    options?: string[];
    points: number;
    order_index: number;
}

export interface QuizAttempt {
    id: number;
    quiz_id: number;
    user_id: string;
    started_at: string;
    completed_at?: string;
    time_taken?: number;
    total_points: number;
    earned_points: number;
    score: number;
    passed: boolean;
    status: "in_progress" | "completed";
    skill_points_awarded: boolean;
    answers?: QuizAnswerResult[];  // Add answers field for detailed attempt data
}

export interface QuizSubmission {
    answers: QuizAnswer[];
}

export interface QuizAnswer {
    question_id: number;
    answer_text: string;
}

export interface QuizResult {
    attempt_id: number;
    quiz_id: number;
    user_id: string;
    score: number;
    total_points: number;
    earned_points: number;
    passed: boolean;
    skill_points_awarded: number;
    feedback: string;
    completed_at: string;
    time_taken: number;
    answers: QuizAnswerResult[];
}

export interface QuizAnswerResult {
    question_id: number;
    question_text: string;
    user_answer: string;
    correct_answer: string;
    is_correct: boolean;
    points_earned: number;
    ai_feedback: string;
    explanation?: string;
}

export interface QuizGenerationRequest {
    module_id: number;
    num_questions: number;
}

export interface QuizGenerationRequest {
    module_id: number;
    num_questions: number;
}

export interface QuizGenerationResponse {
    message: string;
    learning_path_title: string;
    total_modules: number;
    modules_needing_quizzes: number;
    status: "background_task_scheduled" | "no_action_needed";
}

File: be/app/models/quiz.py
from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, JSON, Enum as SQLEnum, Boolean, Text, Float
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from enum import Enum
from datetime import datetime, timezone
from app.db.database import Base


def get_current_time():
    """Return current UTC time for database defaults"""
    return datetime.now(timezone.utc)


class QuestionType(str, Enum):
    MULTIPLE_CHOICE = "multiple_choice"
    TRUE_FALSE = "true_false"
    SHORT_ANSWER = "short_answer"

class QuizStatus(str, Enum):
    DRAFT = "draft"
    ACTIVE = "active"
    COMPLETED = "completed"
    ABANDONED = "abandoned"

class Quiz(Base):
    __tablename__ = 'quizzes'
    
    id = Column(Integer, primary_key=True)
    module_id = Column(Integer, ForeignKey("modules.id", ondelete="CASCADE"), nullable=False)
    title = Column(String(255), nullable=False)
    description = Column(Text)
    total_questions = Column(Integer, default=0)
    passing_score = Column(Float, default=0.7)  # 70% passing threshold
    estimated_completion_time = Column(Integer, default=30)  # Estimated time in minutes
    is_active = Column(Boolean, default=True)
    
    # Relationships
    module = relationship("Module", back_populates="quiz")
    questions = relationship("Question", back_populates="quiz", cascade="all, delete-orphan")
    attempts = relationship("QuizAttempt", back_populates="quiz", cascade="all, delete-orphan")

class Question(Base):
    __tablename__ = 'questions'
    
    id = Column(Integer, primary_key=True)
    quiz_id = Column(Integer, ForeignKey("quizzes.id", ondelete="CASCADE"), nullable=False)
    question_text = Column(Text, nullable=False)
    question_type = Column(SQLEnum(QuestionType), nullable=False)
    options = Column(JSON)  # For multiple choice options
    correct_answer = Column(Text, nullable=False)
    explanation = Column(Text)
    points = Column(Integer, default=1)
    order_index = Column(Integer, default=0)
    
    # Relationships
    quiz = relationship("Quiz", back_populates="questions")
    answers = relationship("Answer", back_populates="question", cascade="all, delete-orphan")

class QuizAttempt(Base):
    __tablename__ = 'quiz_attempts'
    
    id = Column(Integer, primary_key=True)
    quiz_id = Column(Integer, ForeignKey("quizzes.id", ondelete="CASCADE"), nullable=False)
    user_id = Column(String(36), ForeignKey("users.id"), nullable=False)
    status = Column(SQLEnum(QuizStatus), default=QuizStatus.DRAFT)
    score = Column(Float, default=0.0)
    total_points = Column(Integer, default=0)
    earned_points = Column(Integer, default=0)
    passed = Column(Boolean, default=False)
    feedback = Column(Text)  # AI-generated feedback
    skill_points_awarded = Column(Boolean, default=False)  # Track if skill points were awarded for this attempt
    
    # Time tracking fields
    started_at = Column(DateTime(timezone=True), default=get_current_time)
    completed_at = Column(DateTime(timezone=True), nullable=True)
    time_taken = Column(Integer, nullable=True)  # Time taken in seconds
    
    # Relationships
    quiz = relationship("Quiz", back_populates="attempts")
    user = relationship("User", back_populates="quiz_attempts")
    answers = relationship("Answer", back_populates="attempt", cascade="all, delete-orphan")

class Answer(Base):
    __tablename__ = 'answers'
    
    id = Column(Integer, primary_key=True)
    attempt_id = Column(Integer, ForeignKey("quiz_attempts.id", ondelete="CASCADE"), nullable=False)
    question_id = Column(Integer, ForeignKey("questions.id", ondelete="CASCADE"), nullable=False)
    answer_text = Column(Text, nullable=False)
    is_correct = Column(Boolean, default=False)
    points_earned = Column(Integer, default=0)
    ai_feedback = Column(Text)  # AI-generated feedback for the answer
    
    # Relationships
    attempt = relationship("QuizAttempt", back_populates="answers")
    question = relationship("Question", back_populates="answers")
